contrib/win32/win32compat/misc.c

```diff
@@ -1115,216 +1102,165 @@ get_machine_domain_name(wchar_t *domain, int size)
+		goto cleanup;
 	}
 
-	if (user_domain) {
-		/* Fetch Domain groups */
-		nStatus = NetGetDCName(NULL, machine_domain_name_utf16, (LPBYTE *)&dc_name_utf16);
-		if (NERR_Success == nStatus) {
-			debug3("domain controller name: %ls", dc_name_utf16);
-
-			nStatus = NetUserGetLocalGroups(dc_name_utf16,
-				    user_name_utf16,
-				    0,
-				    flags,
-				    (LPBYTE *)&domain_groups,
-				    MAX_PREFERRED_LENGTH,
-				    &num_domain_groups_read,
-				    &total_domain_groups);
-
-			if (NERR_Success != nStatus)
-				error("Failed to get domain groups from DC:%s error: %d\n", dc_name_utf16, nStatus);
-		}
-		else
-			error("Failed to get the domain controller name, error: %d\n", nStatus);
-
-		/* Fetch global, universal groups */
-		nStatus = NetUserGetGroups(dc_name_utf16,
-			user_name_utf16,
-			0,
-			(LPBYTE *)&global_universal_groups,
-			MAX_PREFERRED_LENGTH,
-			&num_global_universal_groups_read,
-			&total_global_universal_groups);
-
-		if (NERR_Success != nStatus)
-			error("Failed to get global,universal groups from DC:%ls error: %d\n", dc_name_utf16, nStatus);
+	/* allocate and copy the sids to a a structure we can pass to lookup */
+	if ((group_sids = (PSID *)malloc(sizeof(PSID) * group_buf->GroupCount)) == NULL) {
+		errno = ENOMEM;
+		goto cleanup;
 	}
 
-	int total_user_groups = num_local_groups_read + num_domain_groups_read + num_global_universal_groups_read;
+	DWORD group_sids_count = 0;
+	for (DWORD i = 0; i < group_buf->GroupCount; i++) {
 
-	/* populate the output */
-	user_groups = malloc(total_user_groups * sizeof(*user_groups));	
+		/* only bother with group thats are 'enabled' from a security perspective */
+		if (group_buf->Groups[i].Attributes & SE_GROUP_ENABLED == 0 ||
+			!IsValidSid(group_buf->Groups[i].Sid))
+			continue;
 
-	populate_user_groups(user_groups, &num_user_groups, num_local_groups_read, total_local_groups, (LPBYTE) local_groups, LOCAL_GROUP);
-	if (user_domain) {
-		populate_user_groups(user_groups, &num_user_groups, num_domain_groups_read, total_domain_groups, (LPBYTE)domain_groups, DOMAIN_GROUP);
-		populate_user_groups(user_groups, &num_user_groups, num_global_universal_groups_read, total_global_universal_groups, (LPBYTE)global_universal_groups, GLOBAL_UNIVERSAL_GROUP);
+		/* only bother with groups that are builtin or classic domain/local groups 
+		 * also ignore domain users and builtin users since these will be meaningless 
+		 * since they do not resolve properly on workgroup computers; these would 
+		 * never meaningfully be used in the server configuration */
+		SID * sid = group_buf->Groups[i].Sid;
+		DWORD sub = sid->SubAuthority[0];
+		DWORD rid = sid->SubAuthority[sid->SubAuthorityCount - 1]; 
+		SID_IDENTIFIER_AUTHORITY nt_authority = SECURITY_NT_AUTHORITY;
+		if (memcmp(&nt_authority, GetSidIdentifierAuthority(sid), sizeof(SID_IDENTIFIER_AUTHORITY)) == 0 && (
+			sub == SECURITY_NT_NON_UNIQUE || sub == SECURITY_BUILTIN_DOMAIN_RID) &&
```

Comment:
Just wanted to note that the check (sub == SECURITY_BUILTIN_DOMAIN_RID) is to eliminate 'None' group from showing up (and causing the AppVeyor tests to fail that check for 0 group memberships).  The 'None' group is a real principal that's added to every local account (sub-authority: '-513').  When the accounts is domain joined, this resolves to 'Domain Users'.  You can actually assign permissions on the file system to 'None' and it's effectively the same as granting it to 'Authenticated Users'. It looks like the author of the 'whoami /groups' command actually tries to exclude this as well which has the side effect of making 'Domain Users' not show up on domain accounts (which it definitely should).  The ‘Users’ BUILTIN group is also always added to the token regardless of the user being in the ‘Users’ group.  Since these are all arguably real group principals that are used in group-based security checks, one could argue we should include them as well.  If we choose to do this though, we’d have to eliminate the “no groups” tests from AppVeyor since there’s really no such thing as a real user with no group principals -- the only way that would happen is S4U fails or name lookup would fails with would be hard to trigger in an E2E test.

Reply:
good to add this info as a comment in code  :)

Reply:
@manojampalam I can certainly expand the comment in the code, but I guess I meant this more as a question to you guys.  Since 'None' is a real group security principal that Windows will honor for other security checks within Windows, maybe we should eliminate this check altogether and include it in the group list?  I think that is the *right* thing to do and it would simplify the code, but I will have to eliminate the E2E tests that check for no groups since there won't be a realistic way to stage that test.  Let me know...